r/src/altrep.cpp

```diff
@@ -104,63 +143,138 @@ struct ArrayNoNull {
+    R_set_altrep_Length_method(class_t, AltrepVector::Length);
+    R_set_altrep_Inspect_method(class_t, AltrepVector::Inspect);
+    R_set_altrep_Duplicate_method(class_t, AltrepVector::Duplicate);
 
     // altvec
-    R_set_altvec_Dataptr_method(class_t, ArrayNoNull::Dataptr);
-    R_set_altvec_Dataptr_or_null_method(class_t, ArrayNoNull::Dataptr_or_null);
+    R_set_altvec_Dataptr_method(class_t, AltrepVector::Dataptr);
+    R_set_altvec_Dataptr_or_null_method(class_t, AltrepVector::Dataptr_or_null);
   }
 };
 
-struct DoubleArrayNoNull {
+struct AltrepVectorDouble {
+  using Base = AltrepVector<REALSXP>;
   static R_altrep_class_t class_t;
 
   static void Init(DllInfo* dll) {
-    class_t = R_make_altreal_class("array_nonull_dbl_vector", "arrow", dll);
-    ArrayNoNull<REALSXP>::Init(class_t, dll);
-    R_set_altreal_No_NA_method(class_t, ArrayNoNull<REALSXP>::No_NA);
+    class_t = R_make_altreal_class("array_dbl_vector", "arrow", dll);
+    Base::Init(class_t, dll);
+
+    R_set_altreal_No_NA_method(class_t, Base::No_NA);
+
+    R_set_altreal_Sum_method(class_t, Base::Sum);
+    R_set_altreal_Min_method(class_t, Base::Min);
+    R_set_altreal_Max_method(class_t, Base::Max);
   }
 
-  static SEXP Make(const std::shared_ptr<Array>& array) {
-    return ArrayNoNull<REALSXP>::Make(class_t, array);
+  static SEXP Make(const std::shared_ptr<Array>& array, RTasks& tasks) {
+    return Base::Make(class_t, array, tasks);
   }
 };
 
-struct Int32ArrayNoNull {
+struct AltrepVectorInt32 {
+  using Base = AltrepVector<INTSXP>;
   static R_altrep_class_t class_t;
 
   static void Init(DllInfo* dll) {
-    class_t = R_make_altinteger_class("array_nonull_int_vector", "arrow", dll);
-    ArrayNoNull<INTSXP>::Init(class_t, dll);
-    R_set_altinteger_No_NA_method(class_t, ArrayNoNull<INTSXP>::No_NA);
+    class_t = R_make_altinteger_class("array_int_vector", "arrow", dll);
+    Base::Init(class_t, dll);
+    R_set_altinteger_No_NA_method(class_t, Base::No_NA);
+
+    R_set_altinteger_Sum_method(class_t, Base::Sum);
+    R_set_altinteger_Min_method(class_t, Base::Min);
+    R_set_altinteger_Max_method(class_t, Base::Max);
   }
 
-  static SEXP Make(const std::shared_ptr<Array>& array) {
-    return ArrayNoNull<INTSXP>::Make(class_t, array);
+  static SEXP Make(const std::shared_ptr<Array>& array, RTasks& tasks) {
+    return Base::Make(class_t, array, tasks);
   }
 };
 
-R_altrep_class_t Int32ArrayNoNull::class_t;
-R_altrep_class_t DoubleArrayNoNull::class_t;
+R_altrep_class_t AltrepVectorInt32::class_t;
+R_altrep_class_t AltrepVectorDouble::class_t;
 
 void Init_Altrep_classes(DllInfo* dll) {
-  DoubleArrayNoNull::Init(dll);
-  Int32ArrayNoNull::Init(dll);
+  AltrepVectorDouble::Init(dll);
+  AltrepVectorInt32::Init(dll);
 }
 
-SEXP MakeDoubleArrayNoNull(const std::shared_ptr<Array>& array) {
-  return DoubleArrayNoNull::Make(array);
+SEXP MakeAltrepVectorDouble(const std::shared_ptr<Array>& array, RTasks& tasks) {
```

Comment:
Since `tasks` is a mutable argument, can you make it `RTasks* tasks`?

Reply:
we no longer use tasks there as there is nothing to parallelise